win, pipe: avoid synchronous pipe deadlocks

bzoz opened the pull request:
Add a thread that will interrupt `uv_pipe_zero_readdile_thread_proc` every half second. This allows other processes to access the pipe without deadlocking. The "interrupt" is done by a call to `CancelSynchronousIo`, similar to how `uv__pipe_getname` pauses the read (see https://github.com/joyent/libuv/issues/1313)

On Windows a call to `uv_pipe_open` will deadlock if another thread or process has `ReadFile` call active on that pipe. 

This can happen when a process spawns Node with `STDIN` redirected to pipe. Libuv will call `ReadFile` in a separate thread ([win/pipe.c:971](https://github.com/libuv/libuv/blob/9d94fd978fc3fa8b32721ce523e46262ab88e392/src/win/pipe.c#L971)). If this Node app spawns another child Node process with inherited stdio handles, this child process will hang. This is the reason behind https://github.com/nodejs/node/issues/10836.

`uv_pipe_open` hangs on a call to `NtQueryInformationFile` ([win/pipe.c:1931](https://github.com/libuv/libuv/blob/9d94fd978fc3fa8b32721ce523e46262ab88e392/src/win/pipe.c#L1931)). This can be solved by using `NtQueryObject`. Unfortunately, this does not solve the issue - a later call to `uv_set_pipe_handle` ([win/pipe.c:245](https://github.com/libuv/libuv/blob/9d94fd978fc3fa8b32721ce523e46262ab88e392/src/win/pipe.c#L245)) uses `SetNamedPipeHandleState`, `GetNamedPipeHandleState` and `NtQueryInformationFile` (with different parameters).

This is related to https://github.com/joyent/libuv/issues/1313. But since the `ReadFile` is used in another process than `uv_open_pipe` we cannot synchronize them.

Ref: https://github.com/joyent/libuv/issues/1313 (similar issue in libuv)
Ref: https://github.com/nodejs/node/issues/10836 (issue in node fixed by this)

bzoz commented:
@saghul could you take a look?

vtjnash commented on src/win/pipe.c:
```diff
@@ -965,9 +985,17 @@ static DWORD WINAPI uv_pipe_zero_readfile_thread_proc(void* parameter) {
 restart_readfile:
+  if (hThreadForInterrupt) {
+    QueueUserWorkItem(uv__pipe_zero_read_interrupter,
+                      hThreadForInterrupt,
```
It's undefined what Win32 will do with this handle after this function returns (I expect it should reuse it for the next IO call). Provision needs to be made to ensure that the work item scheduled here doesn't try to access this handle asynchronously. This is what the mutex locks in `uv__pipe_pause_read ` are required for.

bzoz commented:
I've simplified the logic there. Now the "interrupter-thread" is created once and it calls `CancelSynchronousIo` in a loop. This thread will exit when `uv_pipe_zero_readfile_thread_proc` exits (`WaitForSingleObject` returns something different than `WAIT_TIMEOUT`)

@bnoordhuis @cjihrig could you review this?

@vtjnash I'm sorry, I don't really understand what is the issue here. The duplicated thread handle should be perfectly safe to be used in another thread without any synchronization. The mutex in `uv__pipe_pause_read` looks more like a way to signal `uv_pipe_zero_readfile_thread_proc` to resume the read.

cjihrig commented on test/test-pipe-open-read-pipe.c:
```diff
@@ -0,0 +1,61 @@
+  /* Give uv_run some time to start */
+  uv_sleep(250);
+  /* Try to access the pipe again, in different loop */
+  uv_loop_t test_loop;
```
Can you move this to the top of the function.

cjihrig commented on test/test-pipe-open-read-pipe.c:
```diff
@@ -0,0 +1,61 @@
+}
+
+void pipe_read_thread_proc(void* arg) {
+  uv_pipe_t* pipe = arg;
```
Can you separate the assignment.

cjihrig commented on src/win/pipe.c:
```diff
@@ -942,13 +945,33 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
   uv_pipe_t* handle = (uv_pipe_t*) req->data;
   uv_loop_t* loop = handle->loop;
   HANDLE hThread = NULL;
+  HANDLE hThreadForInterrupt = NULL;
```
I think the assignment should be separate, but clearly the surrounding code isn't doing that.

vtjnash commented:
> The duplicated thread handle should be perfectly safe to be used in another thread without any synchronization

Why do you assume that? I would be surprised if the thread-pool was guaranteed to create a new thread for every work item submitted – that would be a lot more expensive than the mutex I'm suggesting may be necessary to synchronize the shared state between the two threads.

> This can happen when a process spawns Node with STDIN redirected to pipe. Libuv will call ReadFile in a separate thread (win/pipe.c:971). If this Node app spawns another child Node process with inherited stdio handles, this child process will hang. This is the reason behind nodejs/node#10836.

I don't understand why this causes node to hang, since it doesn't look like your example code is reading from STDIN. Is node starting an active read on the STDIN pipe even without an `.on("data")` handler?

bzoz commented:
In [MSDN `DuplicateHandle`](https://msdn.microsoft.com/pl-pl/library/windows/desktop/ms724251(v=vs.85).aspx) example it is used in similar way as here - a handle to object (in that case it is a mutex) is duplicated and passed on to a worker thread. Worker thread then uses the duplicated handle without any synchronization. Here we do exactly the same thing. Handle from `GetCurrentThread()` is duplicated into `hThreadForInterrupt` and passed on to thread-pool job. Worker thread uses this handle exclusively, it is not passed to any other thread or function.

Original node issue with repro using 3 files is here: https://github.com/nodejs/node/issues/10836. My test case boils it down to the `uv_pipe_open` call on an anonymous pipe that is already read by another thread. 

What happens in node is the following:
- node process `A` spawns child node process `B` with `{stdio: 'pipe'}` option
- node process `B` starts reading from stdin - because stdio is a pipe this creates `uv_pipe_zero_readfile_thread_proc` which calls `ReadFile`
- node process `B` spawns child node process `C` with the default `{stdio: inherit}`
- node process `C` access `process.stdin` - this will call `uv_pipe_open` and in turn `NtQueryInformationFile` on a pipe read by `B` and deadlock

vtjnash commented:
Yes, that example is fine because it doesn't use the thread pool and explicitly uses the thread object itself as a mutex. This is valid, but very slow and resource intensive.

> What happens in node is the following:

I was looking at that, but I don't see why node process `B` is reading from `stdin`. I don't see anywhere in the `middle.js` code that is starting a read on `stdin`?

bzoz commented:
Sorry @vtjnash , I'm lost. What are your concerns exactly? I don't see anything particularly wrong with this code, am I missing something?

As for the read: it seems that node starts reading the pipe and then immediately pause it, as described here: https://github.com/nodejs/node/pull/5776 (this was later reverted). In any case the reason for deadlock can be easily verified by attaching debugger to node processes.

bzoz commented:
@cjihrig I've corrected the code.

I've also made the duplicated thread handle not inheritable.

vtjnash commented:
The `uv_pipe_zero_readfile_thread_proc` function is called in a thread-pool (QueueUserWorkItem), so it does not own the thread that it is running on. It is undefined behavior to have any copy of this handle live when the `uv_pipe_zero_readfile_thread_proc` function returns.

> As for the read

This sounds like a bug in nodejs. It shouldn't be initializing and attempting a read when it doesn't want the data (since that wastes resources – and causes the hang observed here).

bzoz commented:
Ah, I see now. I was missing the part that `uv_pipe_zero_readfile_thread_proc` is not a "normal" thread. Thanks for pointing it out, you are right, I'll fix that.

As for node - the linked PR tried to fix that and it caused other problems. OTOH, a process could be reading from stdin and create child processes, and we would still get this bug.

vtjnash commented:
> OTOH, a process could be reading from stdin and create child processes, and we would still get this bug

Not necessarily – that could also be considered a bug in the javascript app that is trying to simultaneously consume stdin from two places. (instead of incurring the overhead here of starting an extra thread and waking up both threads twice every second).

bzoz commented:
I've updated the PR code, PTAL.

vtjnash commented on src/win/pipe.c:
```diff
@@ -942,20 +951,81 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+  uv__pipe_interrupter_param_t* param;
+
+  param = (uv__pipe_interrupter_param_t*) uv__malloc(
+    sizeof(uv__pipe_interrupter_param_t));
```
this should be able to use `alloca` – the application needs to guarantee that this pointer is never accessed after the function returns, to avoid race conditions, anyways.

vtjnash commented on src/win/pipe.c:
```diff
@@ -942,20 +951,81 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+  for (;;) {
+    r = WaitForSingleObject(interrupterParam->completedEvent,
+                            PIPE_ZERO_READ_INTERRUPT_INTERVAL);
+    if (r == WAIT_TIMEOUT) {
```
this is still race-y. getting a wait-timeout previous to here doesn't guarantee that the thread handle is still valid after here.

vtjnash commented on src/win/pipe.c:
```diff
@@ -46,6 +46,9 @@ struct uv__ipc_queue_item_s {
 static char uv_zero_[] = "";
 
+/* To prevent deadlocks we will interrupt synchronous pipe read every 500 ms */
+#define PIPE_ZERO_READ_INTERRUPT_INTERVAL 500
```
This seems very frequent - especially for something that will likely be caused by stdio, having timers that need to get woken up at a 500ms period seems like it could have a measurable impact on battery life. For non-nodejs consumers of this library (i.e. programs that don't share the nodejs bug, like my use case of JuliaLang), there should also be a way to disable this code.

bzoz commented:
@vtjnash I've updated the code. I've added a mutex for synchronization, bumped interval to 2.5s and made entire thing opt-in by a call to `uv_pipe_enable_interrupter`.

bzoz commented:
ping?

bzoz commented:
pong?

bzoz commented:
ping?

bnoordhuis commented:
Sorry Bartosz, I missed that you asked for a review.  I've only looked at this briefly but it seems like a rather inelegant solution.  What would happen if `uv_pipe_open()` duplicates the handle before using it?

bzoz commented:
I guess this would affect both handles. I'll double check.

bzoz commented:
@bnoordhuis each `uv_pipe_t` instance needs a separate call to `uv_pipe_enable_interrupter`. 

I know this is not an elegant solution, and as such per @vtjnash suggestion it is on-demand. But this is the only way I know of solving this, which is a real-world issue.

Trott commented:
Let's try to loop in a few more folks for some reviews on this...

@nodejs/platform-windows

Trott commented:
Whoops, wrong repo. Sorry.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -76,6 +79,13 @@ typedef struct {
+typedef struct {
+  volatile HANDLE thread;
+  HANDLE threadMutex;
+  HANDLE completedEvent;
```
Minor stylistic issue: snake_case, not camelCase.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+  uv__readfile_interrupter_t* interrupter;
+  int r;
+
+  interrupter = (uv__readfile_interrupter_t*) param;
```
Superfluous cast?

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+      }
+      ReleaseMutex(interrupter->threadMutex);
+    } else {
+      /* ReadFile thread has terminated - cleanup the handles and exit */
```
Typo: clean up

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+  uv__readfile_interrupter_t* interrupter;
+
+  interrupter = (uv__readfile_interrupter_t*) uv__malloc(
+    sizeof(uv__readfile_interrupter_t));
```
`sizeof(*interrupter)`.  Also, superfluous cast.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+}
+
+static uv__readfile_interrupter_t* uv__start_readfile_interrupter() {
+  HANDLE threadHandle;
```
Style: snake_case, not camelCase.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+    goto failed_event;
+  }
+  if (!QueueUserWorkItem(uv__readfile_interrupter_thread, interrupter,
+    WT_EXECUTEINLONGTHREAD)) {
```
Can you line up the argument?

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+  interrupter = (uv__readfile_interrupter_t*) uv__malloc(
+    sizeof(uv__readfile_interrupter_t));
+  if (interrupter == NULL) {
+    goto failed_malloc;
```
I'd just `return NULL` here.  No need to jump, nothing to clean up.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+    WT_EXECUTEINLONGTHREAD)) {
+    goto failed_queue;
+  }
+  goto done;
```
Just `return interrupter`?

bnoordhuis commented on test/test-pipe-open-read-pipe.c:
```diff
@@ -0,0 +1,63 @@
+  HANDLE stdin_read_pipe, stdin_write_pipe;
+  SECURITY_ATTRIBUTES sa_attr;
+
+  sa_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
```
`sizeof(sa_attr)`?

bzoz commented:
I know. It is a hack, [it uses hackish things](https://github.com/libuv/libuv/pull/1273/files#diff-7be27b71e7f7a6d2c40c447747db1231R1083), but it works: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/375/

saghul commented:
I'll take a look.

vtjnash commented:
I agree that it's still a bit strange to work around a nodejs bug in the libuv layer, but this approach should work.

saghul commented:
I took a quick look... and it's nasty 😞  Not your fault @bzoz!

Is there a reason why we don't do this always for sync pipes? I mean, who wouldn't want to avoid deadlocks? Specially given users don't know if a pipe is synchronous or not, right?

vtjnash commented:
The operation is a no-op for asynchronous pipes (they don't have the same issue in the kernel). I requested that it not be always enable for synchronous pipes since this is only needed as a workaround for a nodejs bug. Other frontends (for example, my use case is Julia), don't need the extra overhead of spawning and synchronizing yet another thread.

saghul commented:
Can you clarify why it's not needed? If it's a kernel thing, why isn't Julia running into the same problem?

vtjnash commented:
Julia's standard library doesn't execute a read on a pipe if there's no active reader at the language / user level. That ensures it's only an issue in edge cases (two readers in separate processes trying to consume tokens off of the same pipe).

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+}
+
+static uv__readfile_interrupter_t* uv__start_readfile_interrupter() {
+  HANDLE thread_handle;
```
() -> (void) - () is a function taking any number of arguments.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -985,6 +1079,8 @@ static DWORD WINAPI uv_pipe_zero_readfile_thread_proc(void* parameter) {
           handle->pipe.conn.readfile_thread = hThread;
           uv_mutex_unlock(m);
+          /* Give some time for other processes to wake before restarting. */
+          Sleep(1);
```
Is this necessary?  What happens if you remove it?

bzoz replied:
Unfortunately - yes. Without it the test won't work. Interrupter thread will be stuck at [CancelSynchronousIo](https://github.com/libuv/libuv/pull/1273/files#diff-7be27b71e7f7a6d2c40c447747db1231R968), zero-read at [ReadFile](https://github.com/libuv/libuv/pull/1273/files#diff-7be27b71e7f7a6d2c40c447747db1231R1066), and the test thread at [NtQueryInformationFile](https://github.com/libuv/libuv/pull/1273/files#diff-7be27b71e7f7a6d2c40c447747db1231R2037). For whatever reason zero-read thread seems to ignore the `CancelSynchronousIo` call, unless a `Sleep` (or a `printf`, or a attached debugger) is present.

vtjnash replied:
This should be conditional on `interrupter`.

bnoordhuis commented on src/win/pipe.c:
```diff
@@ -1010,6 +1106,9 @@ static DWORD WINAPI uv_pipe_zero_readfile_thread_proc(void* parameter) {
   }
 
   POST_COMPLETION_FOR_REQ(loop, req);
+  if (interrupter) {
```
`if (interrupter != NULL) {`

bnoordhuis commented on test/test-pipe-open-read-pipe.c:
```diff
@@ -0,0 +1,81 @@
+  ASSERT(r == 0);
+
+  /* Give uv_run some time to start */
+  uv_sleep(250);
```
I hope this isn't going to be flaky...

bzoz commented:
@bnoordhuis updated, PTAL

vtjnash commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+        SwitchToThread();
+      }
+      ReleaseMutex(interrupter->thread_mutex);
+    } else {
```
`if (WAIT_OBJECT_0)`

bzoz commented:
@vtjnash updated, PTAL

bzoz commented:
If no one objects, I will land this tomorrow.

saghul commented:
I still don't like that one function that is super-specific. Also pretty much Node only AFAWCT. Can we wake this behavior a compile-time define? We can always turn it into a function if it's needed later, but I'd really want to avoid adding this weird to use API.

vtjnash commented:
> Can we wake this behavior a compile-time define?

I thought so far that libuv has been avoiding compile-time defines? This one seems like it would be especially inconvenient for system packagers (if there are any on Windows?), as it forces them to choose between bad latency (read monitoring is only active 70% of the time with this PR) for all applications or nodejs/node#10836.

vtjnash commented:
See also https://github.com/libuv/libuv/issues/95#issuecomment-68413692, which this PR implements.

bzoz commented:
Using `WaitForSingleObject` was the first thing I tried, since I basically want to add a timeout to the `ReadFile`. I was not able to make it work, it looks like one cannot synchronize on pipes ([see MSDN](https://msdn.microsoft.com/pl-pl/library/windows/desktop/ms686364.aspx)).

@vtjnash I've added a slight break which happens every 2.5s. Does this really end in the read being active only 70% of the time?

I think that this API is suboptimal. Basically each synchronous pipe will be affected by this bug and can cause a deadlock. I think that all synchronous pipes read should have interrupter enabled. I will drop `uv_pipe_enable_interrupter` and make it opt-in through an option to `uv_loop_configure`.

bzoz commented on src/win/pipe.c:
```diff
@@ -942,20 +952,101 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
+
+  if (!QueueUserWorkItem(uv__readfile_interrupter_thread,
+                         interrupter,
+                         WT_EXECUTEINLONGTHREAD))
```
[MSDN](https://msdn.microsoft.com/pl-pl/library/windows/desktop/ms685061.aspx) says I can't wait on a pipe.

I could use [CreateTimerQueueTimer](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682485.aspx) to use the threadpool to wake interrupter every 2.5s, but I think that would complicate things.

vtjnash replied:
Sadly true. I spent several hours yesterday trying to work around the NT kernel behavior (and failing), and wishing that it was open source.

But that's not why you're allocating this work item. This thread call can be replaced with a `RegisterWaitForSingleObject` callback.

bzoz replied:
On what handle should I wait?

vtjnash replied:
`interrupter->completed_event`

bzoz commented:
So, adding an option to `uv_loop_configure` won't work, because there is no `flags` or anything similar in `uv_loop_t` on Windows ([[1]](https://github.com/libuv/libuv/blob/564677d325db6122cba2091100d07ea6d366161b/include/uv.h#L1487), [[2]](https://github.com/libuv/libuv/blob/564677d325db6122cba2091100d07ea6d366161b/include/uv-win.h#L312)).

So again - while this is not a nice API, and implementation is basically a hack, I don't see any other option.

BTW, this has been open for 5 months now 🎂

vtjnash commented:
>  I don't see any other option.

Got any twitter followers on the NT Kernel team? :) I'm still hoping that someday that they'll fix this bug.

Is it a competition? I've got https://github.com/libuv/libuv/pull/1498 with 5 years and counting 😛. You are far more active around here than me though.

saghul commented:
> I thought so far that libuv has been avoiding compile-time defines?

Special problems call for special measures.

> BTW, this has been open for 5 months now 🎂

It has been broken for a lot longer though 🤷‍♂️

bzoz commented:
Since I cannot add this to `uv_loop_configure`, maybe we should add a global configuration (eg. `uv_configure`)? Besides opting-in for this, it could maybe be used to opt-out from user-mode symlinks (https://github.com/libuv/libuv/pull/1466) ?

saghul commented:
> Besides opting-in for this, it could maybe be used to opt-out from user-mode symlinks (#1466) ?

Why would anyone ever want that?

saghul commented:
My stance on this, unless there is a new approach which we haven't seen yet is to do it compile time.

bzoz commented:
The user-mode symlinks thing was only an example.

I'm reluctant to make this compile time. Mostly I think because there is no such mechanism already in place. IMHO a call to `uv_configure` (or `uv_loop_configure` with `NULL` as loop parameter) would be nicer and easier to maintain than a new build feature.

BTW, @vtjnash are you sure it should not be default in libuv? Even though in this specific situation is caused by the way Node handles stdio, Julia app could receive a pipe that is already being read as one of its stdio handles. Trying to initialize such pipe with `uv_pipe_open` will deadlock any runtime.

vtjnash commented:
I don't think that protecting applications against deadlock caused by other misbehaving applications is really in the purview of libuv. For more context, this was also discussed on the original PR (https://github.com/joyent/libuv/pull/1377#issuecomment-51681657). I'll admit that it's unsatisfying that the deadlock could be avoided if Microsoft fixed their kernel. However it should be rare to encounter situations that it would occur (an application that spawns nodejs after spawning ReadFile in a thread pool on the same NamedPipe it passes to nodejs to use as stdin – note that this situation also creates an race condition of two processes trying to consume user keystroke input).

bzoz commented:
As compile time option: https://github.com/libuv/libuv/pull/1525

vtjnash commented:
> Trying to initialize such pipe with uv_pipe_open will deadlock any runtime

This PR does not address this case. This PR can only help with working around nodejs bugs, there is _nothing_ that can be done in libuv that would work around bugs in other applications.
